prov/verbs/src/verbs_cq.c

```diff
@@ -297,32 +297,37 @@ int vrb_save_wc(struct vrb_cq *cq, struct ibv_wc *wc)
 	return FI_SUCCESS;
 }
 
-static void vrb_flush_cq(struct vrb_cq *cq)
+static void vrb_flush_cq(struct vrb_cq *cq, size_t rx_size)
 {
 	struct ibv_wc wc;
-	ssize_t ret;
+	size_t count = 0;
+	int retry = 0;
 
-	ofi_genlock_lock(&cq->util_cq.cq_lock);
-	while (1) {
-		ret = vrb_poll_cq(cq, &wc);
-		if (ret <= 0)
-			break;
+	/* Some HW is populating CQEs in chunks of 64.
+	 * Set max_retry to 2 * rx_size/64 + 1
+	 */
+	int max_retry = rx_size/32 + 1;
```

Comment:
Please declare variables at the top of all functions.  If you need a comment, do that and the assignment after the declaration.  In this case, the comment doesn't match the code exactly, which looks off on first glance.  And the reasoning for how max_retry is determined is questionable.  The code is assuming that the HW will write 64 entries to the CQ every 1 us, kind of.

That a sleep was added to the code is a strong indicator that the code is racy and the sleep simply helps us to get luckier.  (And making it far more difficult to debug later if we hit into races.)  If we need to flush all posted receive operations and the HW isn't doing that in a way that usable, then we should simply track the receives in SW and flush them manually after the CQ has been emptied (after the QP has been destroyed).  That's guaranteed to work without spin sleeping and hoping the HW writes the completions under load within our timing window.